src/token-metadata/tokens-contract-handler.ts

```diff
@@ -0,0 +1,481 @@
+      );
+    }
+  }
+
+  async start() {
+    logger.info(
+      `[token-metadata] found ${
+        this.tokenKind === 'ft' ? 'sip-010-ft-standard' : 'sip-009-nft-standard'
+      } compliant contract ${this.contractId} in tx ${this.txId}, begin retrieving metadata...`
+    );
+    const sw = stopwatch();
+    // This try/catch block will catch any and all errors that are generated while processing metadata
+    // (contract call errors, parse errors, timeouts, etc.). Fortunately, each of them were previously tagged
+    // as retryable or not retryable so we'll make a decision here about what to do in each case.
+    // If we choose to retry, this queue entry will simply not be marked as `processed = true` so it can be
+    // picked up by the `TokensProcessorQueue` at a later time.
+    let processingFinished = false;
+    try {
+      if (this.tokenKind === 'ft') {
+        await this.handleFtContract();
+      } else if (this.tokenKind === 'nft') {
+        await this.handleNftContract();
+      }
+      processingFinished = true;
+    } catch (error) {
+      if (error instanceof RetryableTokenMetadataError) {
+        const retries = await this.db.increaseTokenMetadataQueueEntryRetryCount(this.dbQueueId);
+        if (
+          getTokenMetadataProcessingMode() === TokenMetadataProcessingMode.strict ||
+          retries <= getTokenMetadataMaxRetries()
+        ) {
+          logger.info(
+            `[token-metadata] a recoverable error happened while processing ${this.contractId}, trying again later: ${error}`
+          );
+        } else {
+          logger.warn(
+            `[token-metadata] max retries reached while processing ${this.contractId}, giving up: ${error}`
+          );
+          processingFinished = true;
+        }
+      } else {
+        // Something more serious happened, mark this contract as done.
+        processingFinished = true;
+      }
+    } finally {
+      if (processingFinished) {
+        await this.db.updateProcessedTokenMetadataQueueEntry(this.dbQueueId);
+        logger.info(
+          `[token-metadata] finished processing ${this.contractId} in ${sw.getElapsed()} ms`
+        );
+      }
+    }
+  }
+
+  /**
+   * fetch Fungible contract metadata
+   */
+  private async handleFtContract() {
+    const contractCallName = await this.readStringFromContract('get-name');
+    const contractCallUri = await this.readStringFromContract('get-token-uri');
+    const contractCallSymbol = await this.readStringFromContract('get-symbol');
+
+    let contractCallDecimals: number | undefined;
+    const decimalsResult = await this.readUIntFromContract('get-decimals');
+    if (decimalsResult) {
+      contractCallDecimals = Number(decimalsResult.toString());
+    }
+
+    let metadata: FtTokenMetadata | undefined;
+    if (contractCallUri) {
+      metadata = await this.getMetadataFromUri<FtTokenMetadata>(contractCallUri);
```

Comment:
I'm wondering what to do for conditions where this external URL cannot be fetched/parsed _but_ the above contract-calls to get `decimals`, `symbol`, etc succeeded. 

You could imagine the third-party server for this metadata being unavailable for an extended period of time or even go offline indefinitely, while most apps could still make do with the information from the contract-call (primarily `decimals`).

Reply:
Great point... IMO for now (this PR) we should keep retrying but we could consider later adding another retry limit just for external metadata that is also enforced during strict mode.

Reply:
I'd hazard a guess that there's probably already cases of a 3rd party urls for this already permanently unavailable, and if these are considered retriable errors, then it seems like it could result in two critical situations:
1) The _more_ critical data and already resolved data from the contract-call, like `decimals`, will never be written to the db since the 3rd party url fetch would cause the function to exit here before finishing.
2) If `TOKEN_METADATA_PARSING_CONCURRENCY_LIMIT` (5) or more of these exist, the queue could be stuck indefinitely trying to process them, possibly never getting to process others.

I agree we don't have to solve this entirely in this PR, but I think we should try to prevent either of the above. It could just be something like `metadata = await this.getMetadataFromUri<FtTokenMetadata>(contractCallUri)` is wrapped in a for loop where it can retry fetching a few times, but once it fails those, it is A) _not_ considered retriable, and B) it lets the function continue (leaving `metadata` as undefined).

The only real downside of this approach would be that `image` urls would not be available if the 3rd party url is unresponsive for the duration of the fixed set of retries -- but the fault for that is much more attributable to the 3rd party than us.

Reply:
We could even fill in one of the "image failed to load" placeholders from internet explorer to indicate this kind of failure 😆 

<img width="246" alt="image" src="https://user-images.githubusercontent.com/1447546/168339273-63c88c6a-58e6-4a00-92e4-edcdb4c6218c.png">

Reply:
🤣 lmao that would be great.

I think you're absolutely right though, I didn't consider the queue being permanently full from these dead URLs being retried forever.

Your idea for a for loop sounds great, I'll add that to this PR.